packages/firestore/src/core/sync_engine.ts

```diff
@@ -265,18 +271,33 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
   }
 
   /**
-   * Reconcile the list of synced documents in the local views with those from
-   * persistence.
+   * Reconcile the list of synced documents in an existing view with those
+   * from persistence.
    */
   // PORTING NOTE: Multi-tab only.
-  private async synchronizeLocalView(targetId: TargetId): Promise<void> {
-    return this.localStore
-      .remoteDocumentKeys(targetId)
-      .then(async remoteKeys => {
-        const queryView = this.queryViewsByTarget[targetId];
-        assert(!!queryView, 'Expected queryView to be defined');
-        queryView.view.synchronizeWithRemoteKeys(remoteKeys);
-      });
+  private synchronizeViewAndComputeSnapshot(
+    query: Query,
+    targetId: TargetId
+  ): Promise<ViewChange> {
+    return this.localStore.executeQuery(query).then(async docs => {
+      return this.localStore
+        .remoteDocumentKeys(targetId)
+        .then(async remoteKeys => {
+          const queryView = this.queryViewsByTarget[targetId];
+          assert(!!queryView, 'Cannot reconcile missing view');
+          const viewChange = queryView.view.synchronizeWithPersistedState(
+            docs,
+            remoteKeys,
+            /* resetLimboDocuments= */ this.isPrimary,
+            /* resetCurrent= */ this.isPrimary
```

Comment:
I simplified this a little to unconditionally re-compute the list of limbo documents. In clients that are in-sync this is a no-op, and in clients that are out of sync, this will bring them back into sync. The only bit that I still need to provide is whether to set `current` to false. I added some documentation to clear this up a little.

Notice that I also added a TODO to mark the queries as non-current in SharedClientState. That's probably a three line change but I wanted to keep this PR from exploding even further.